src/pybind/mgr/dashboard/frontend/src/app/shared/components/form-modal/form-modal.component.ts

```diff
@@ -23,32 +20,90 @@ export class FormModalComponent implements OnInit {
   // Input
   titleText: string;
   message: string;
-  fields: CdFormFieldConfig[];
+  fields: CdFormModalFieldConfig[];
   submitButtonText: string;
   onSubmit: Function;
 
   // Internal
-  formGroup: FormGroup;
+  formGroup: CdFormGroup;
+
+  constructor(
+    public bsModalRef: BsModalRef,
+    private formBuilder: CdFormBuilder,
+    private formatter: FormatterService,
+    private dimlessBinaryPipe: DimlessBinaryPipe,
+    private i18n: I18n
+  ) {}
 
-  constructor(public bsModalRef: BsModalRef, private formBuilder: CdFormBuilder) {}
+  ngOnInit() {
+    this.createForm();
+  }
 
   createForm() {
     const controlsConfig = {};
     this.fields.forEach((field) => {
-      const validators = [];
-      if (_.isBoolean(field.required) && field.required) {
-        validators.push(Validators.required);
-      }
-      controlsConfig[field.name] = new FormControl(_.defaultTo(field.value, null), { validators });
+      controlsConfig[field.name] = this.createFormControl(field);
     });
     this.formGroup = this.formBuilder.group(controlsConfig);
   }
 
-  ngOnInit() {
-    this.createForm();
+  private createFormControl(field: CdFormModalFieldConfig): FormControl {
+    let validators: ValidatorFn[] = [];
+    if (_.isBoolean(field.required) && field.required) {
+      validators.push(Validators.required);
+    }
+    if (field.validators) {
+      validators = validators.concat(field.validators);
+    }
+    return new FormControl(
+      _.defaultTo(
+        field.type === 'binary' ? this.dimlessBinaryPipe.transform(field.value) : field.value,
+        null
+      ),
+      { validators }
+    );
+  }
+
+  getError(field: CdFormModalFieldConfig): string {
+    const formErrors = this.formGroup.get(field.name).errors;
+    const errors = Object.keys(formErrors).map((key) => {
+      return this.getErrorMessage(key, formErrors[key], field.errors);
+    });
+    return errors.join('<br>');
+  }
+
+  private getErrorMessage(
+    error: string,
+    errorContext: any,
```

Comment:
It seems as if errorContext is only used as a function and could therefore become a function as type.

Reply:
`errorContext` is the attribute error that is provided by any error that occurs. We only know for sure that it is a function in case of 'binrayMin' and 'binaryMax' as described in L87 and L88.

Errors can have any type as context to them self, that's why `any` is used here.